File: backend/src/main/java/nextpos/app/nextpos/model/entity/Promotion.java
package nextpos.app.nextpos.model.entity;

import jakarta.persistence.*;
import lombok.*;
import nextpos.app.nextpos.model.enums.DiscountType;
import nextpos.app.nextpos.model.enums.PromotionType;
import org.springframework.data.annotation.CreatedBy;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.LastModifiedBy;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

import java.math.BigDecimal;
import java.time.LocalDateTime;

@Entity
@Table(name = "promotions", indexes = {
        @Index(name = "idx_promotion_company", columnList = "company_id"),
        @Index(name = "idx_promotion_code", columnList = "code"),
        @Index(name = "idx_promotion_active_dates", columnList = "is_active, start_date, end_date"),
        @Index(name = "idx_promotion_warehouse", columnList = "warehouse_id"),
        @Index(name = "idx_promotion_type", columnList = "type")
})
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
@EntityListeners(AuditingEntityListener.class)
public class Promotion {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, length = 100)
    private String name;

    @Column(length = 50, unique = true)
    private String code;

    @Column(length = 500)
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(name = "type", nullable = false, length = 30)
    private PromotionType type;

    @Enumerated(EnumType.STRING)
    @Column(name = "discount_type", nullable = false, length = 20)
    private DiscountType discountType;

    @Column(name = "discount_value", precision = 12, scale = 2)
    private BigDecimal discountValue;

    @Column(name = "max_discount_amount", precision = 12, scale = 2)
    private BigDecimal maxDiscountAmount;

    @Column(name = "min_order_amount", precision = 12, scale = 2)
    private BigDecimal minOrderAmount;

    @Column(name = "max_order_amount", precision = 12, scale = 2)
    private BigDecimal maxOrderAmount;

    @Column(name = "usage_limit")
    private Integer usageLimit;

    @Column(name = "usage_limit_per_customer")
    private Integer usageLimitPerCustomer;

    @Column(name = "start_date", nullable = false)
    private LocalDateTime startDate;

    @Column(name = "end_date")
    private LocalDateTime endDate;

    @Column(name = "is_active", nullable = false)
    @Builder.Default
    private Boolean isActive = true;

    @Column(name = "stacking_priority")
    private Integer stackingPriority;

    @Enumerated(EnumType.STRING)
    @Column(name = "stacking_strategy", nullable = false)
    @Builder.Default
    private PromotionStackingStrategy stackingStrategy = PromotionStackingStrategy.BEST_DISCOUNT;

    @Column(name = "buy_quantity")
    private Integer buyQuantity;

    @Column(name = "get_quantity")
    private Integer getQuantity;

    @Column(name = "get_discount_percent")
    private BigDecimal getDiscountPercent;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "buy_product_id")
    private Product buyProduct;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "get_product_id")
    private Product getProduct;

    @Column(name = "company_id", nullable = false, updatable = false)
    private Long companyId;

    @Column(name = "warehouse_id")
    private Long warehouseId;

    @CreatedBy
    @Column(name = "created_by", updatable = false)
    private Long createdBy;

    @CreatedDate
    @Column(name = "created_at", updatable = false, nullable = false)
    private LocalDateTime createdAt;

    @LastModifiedBy
    @Column(name = "updated_by")
    private Long updatedBy;

    @LastModifiedDate
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        if (createdAt == null)
            createdAt = LocalDateTime.now();
        if (updatedAt == null)
            updatedAt = LocalDateTime.now();
        if (isActive == null)
            isActive = true;
    }

    @PreUpdate
    protected void onUpdate() {
        updatedAt = LocalDateTime.now();
    }
}

File: frontend/src/types/promotion.ts
import { PromotionType } from "@/enums/promotionType";
import { DiscountType } from "@/enums/discountType";
import { PromotionStackingStrategy } from "@/enums/promotionStackingStrategy";
import { CustomerGroup } from "@/enums/CustomerGroup";

export interface PromotionResponse {
  id: number;
  name: string;
  code?: string;
  description?: string;
  type: PromotionType;
  discountType: DiscountType;
  discountValue?: number;
  maxDiscountAmount?: number;
  minOrderAmount?: number;
  maxOrderAmount?: number;
  usageLimit?: number;
  usageLimitPerCustomer?: number;
  startDate: string;
  endDate?: string;
  isActive: boolean;
  stackingPriority?: number;
  stackingStrategy: PromotionStackingStrategy;
  buyQuantity?: number;
  getQuantity?: number;
  getDiscountPercent?: number;
  buyProductId?: number;
  getProductId?: number;
  warehouseId?: number;
  companyId: number;
  createdBy: number;
  createdAt: string;
  updatedBy?: number;
  updatedAt?: string;
  productIds: number[];
  categoryIds: number[];
  customerGroups: CustomerGroup[];
}

export interface CreatePromotionRequest {
  name: string;
  code?: string;
  description?: string;
  type: PromotionType;
  discountType: DiscountType;
  discountValue?: number;
  maxDiscountAmount?: number;
  minOrderAmount?: number;
  maxOrderAmount?: number;
  usageLimit?: number;
  usageLimitPerCustomer?: number;
  startDate: string;
  endDate?: string;
  stackingPriority?: number;
  stackingStrategy: PromotionStackingStrategy;
  buyQuantity?: number;
  getQuantity?: number;
  getDiscountPercent?: number;
  buyProductId?: number;
  getProductId?: number;
  warehouseId?: number;
  productIds: number[];
  categoryIds: number[];
  customerGroups: CustomerGroup[];
}

export type UpdatePromotionRequest = Partial<CreatePromotionRequest>;

export interface CartItemDto {
  productId: number;
  quantity: number;
  unitPrice: number;
}

export interface CouponValidationRequest {
  couponCode: string;
  customerId?: number;
  warehouseId: number;
  companyId: number;
  currencyCode: string;
  items: CartItemDto[];
  shippingCost: number;
}

export interface AppliedProductDiscount {
  productId: number;
  discountAmount: number;
  description: string;
}

export interface CouponValidationResponse {
  valid: boolean;
  message: string;
  discountAmount: number;
  discountType: DiscountType;
  appliedDiscountValue: number;
  appliedPromotionId: number;
  appliedPromotionName: string;
  productDiscounts: AppliedProductDiscount[];
  freeShipping: boolean;
}
